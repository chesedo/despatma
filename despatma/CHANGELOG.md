# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.1.1](https://github.com/chesedo/despatma/compare/despatma-v0.1.0...despatma-v0.1.1) - 2024-08-09

### Other
- workspace versioning ([#4](https://github.com/chesedo/despatma/pull/4))
- Update dependencies ([#2](https://github.com/chesedo/despatma/pull/2))
- update failing tests
- force version downgrades
