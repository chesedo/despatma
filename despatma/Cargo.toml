[package]
name = "despatma"
version = "0.1.0"
authors = ["chesedo <pieter@chesedo.me>"]
edition = "2018"
description = "Design Pattern Macro like Loki"
repository = "https://github.com/chesedo/despatma"
license = "MIT"
keywords = ["macro", "design", "patterns"]

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[lib]
proc-macro = true

[dependencies]
despatma-lib = { path = "../despatma-lib", version = "0.1.0" }

proc-macro2 = "1.0"
quote = "1.0"
syn = "1.0"
tokenstream2-tmpl = "=0.1.1"


[dev-dependencies]
despatma-test-helpers = { path = "../despatma-test-helpers" }
despatma-lib = { path = "../despatma-lib", features = ["extra-traits"] }
macrotest = "1.0"
pretty_assertions = "0.6"
syn = { version = "1.0", features = ["extra-traits"] }
trybuild = "1.0"
