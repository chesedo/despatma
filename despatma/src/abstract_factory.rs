use proc_macro2::TokenStream;
use quote::quote;
use syn::parse::{Parse, ParseStream, Result};
use syn::punctuated::Punctuated;
use syn::{parse_quote, ItemTrait, Token, Type, TypeParamBound};

/// Holds the tokens for the attributes passed to an AbstractFactory attribute macro
/// Expects input in the following format
/// ```text
/// some_factory_method_trait, type_1, type_2, ... , type_n
/// ```
///
/// `some_factory_method_trait` needs to be created by the client.
/// It should have one generic type.
/// Every `type_1` ... `type_n` will be filled into this generic type.
#[cfg_attr(test, derive(Eq, PartialEq, Debug))]
pub struct AbstractFactoryAttribute {
    factory_trait: Type,
    sep: Token![,],
    types: Punctuated<Type, Token![,]>,
}

/// Make AbstractFactoryAttribute parsable
impl Parse for AbstractFactoryAttribute {
    fn parse(input: ParseStream) -> Result<Self> {
        Ok(AbstractFactoryAttribute {
            factory_trait: input.parse()?,
            sep: input.parse()?,
            types: input.parse_terminated(Type::parse)?,
        })
    }
}

impl AbstractFactoryAttribute {
    /// Add factory super traits to an `ItemTrait` to turn the `ItemTrait` into an Abstract Factory
    pub fn expand(&self, input_trait: &mut ItemTrait) -> TokenStream {
        // Build all the super traits
        let factory_traits: Punctuated<TypeParamBound, Token![+]> = {
            let types = self.types.iter();
            let factory_name = &self.factory_trait;

            parse_quote! {
                #(#factory_name<#types>)+*
            }
        };

        // Append extra factory super traits
        input_trait.supertraits.extend(factory_traits);

        quote! {
            #input_trait
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use despatma_test_helpers::reformat;
    use syn::parse_str;

    type Result = std::result::Result<(), Box<dyn std::error::Error>>;

    mod abstract_factory {
        use super::*;
        use pretty_assertions::assert_eq;

        #[test]
        fn parse() -> Result {
            let actual: AbstractFactoryAttribute = parse_str("Factory, u32, i64")?;
            let mut expected_types = Punctuated::new();

            expected_types.push(parse_str("u32")?);
            expected_types.push(parse_str("i64")?);

            assert_eq!(
                actual,
                AbstractFactoryAttribute {
                    factory_trait: parse_str("Factory")?,
                    sep: Default::default(),
                    types: expected_types,
                }
            );

            Ok(())
        }

        #[test]
        #[should_panic(expected = "expected `,`")]
        fn missing_types() {
            parse_str::<AbstractFactoryAttribute>("Factory").unwrap();
        }

        #[test]
        fn expand() -> Result {
            let mut t = parse_str::<ItemTrait>("pub trait Abstraction<T>: Display + Extend<T> {}")?;
            let mut input_types = Punctuated::new();

            input_types.push(parse_str("u32")?);
            input_types.push(parse_str("i64")?);

            let actual = &AbstractFactoryAttribute {
                factory_trait: parse_str("Factory")?,
                sep: Default::default(),
                types: input_types,
            }
            .expand(&mut t);

            assert_eq!(
                reformat(&actual),
                "pub trait Abstraction<T>: Display + Extend<T> + Factory<u32> + Factory<i64> {}\n"
            );

            Ok(())
        }
    }
}
