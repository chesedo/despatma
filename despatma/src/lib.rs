mod abstract_factory;
mod visitor;

extern crate proc_macro;

use proc_macro::TokenStream;
use syn::punctuated::Punctuated;
use syn::{parse_macro_input, ItemTrait, Token};
use tokenstream2_tmpl::Interpolate;

use abstract_factory::AbstractFactoryAttribute;
use despatma_lib::TraitSpecifier;
use visitor::VisitorFunction;

#[proc_macro_attribute]
pub fn abstract_factory(tokens: TokenStream, trait_expr: TokenStream) -> TokenStream {
    let mut input = parse_macro_input!(trait_expr as ItemTrait);
    let attributes = parse_macro_input!(tokens as AbstractFactoryAttribute);

    attributes.expand(&mut input).into()
}

#[proc_macro_attribute]
pub fn interpolate_traits(tokens: TokenStream, concrete_impl: TokenStream) -> TokenStream {
    let attributes =
        parse_macro_input!(tokens with Punctuated::<TraitSpecifier, Token![,]>::parse_terminated);

    attributes.interpolate(concrete_impl.into()).into()
}

#[proc_macro]
pub fn visitor(tokens: TokenStream) -> TokenStream {
    let input = parse_macro_input!(tokens as VisitorFunction);

    input.expand().into()
}
