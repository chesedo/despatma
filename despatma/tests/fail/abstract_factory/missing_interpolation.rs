/// Test when concrete trait interpolations are missing
mod library;

use despatma::abstract_factory;
use library::elements::{Button, Element, Window};

// Factory for a single element
pub trait Factory<T: Element + ?Sized> {
    fn create(&self) -> Box<T>;
}

// Abstract Factory for multiple elements
#[abstract_factory(Factory, dyn Button, Window)]
pub trait AbstractGuiFactory {}

// Concrete factory implementing the abstract factory
struct GnomeFactory {}
impl AbstractGuiFactory for GnomeFactory {}

fn main() {}
