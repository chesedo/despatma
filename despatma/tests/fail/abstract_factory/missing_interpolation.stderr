error[E0277]: the trait bound `GnomeFactory: Factory<lib::elements::Window>` is not satisfied
  --> $DIR/missing_interpolation.rs:18:6
   |
13 | #[abstract_factory(Factory, dyn Button, Window)]
   | ------------------------------------------------ required by this bound in `AbstractGuiFactory`
14 | pub trait AbstractGuiFactory {}
   |           ------------------ required by a bound in this
...
18 | impl AbstractGuiFactory for GnomeFactory {}
   |      ^^^^^^^^^^^^^^^^^^ the trait `Factory<lib::elements::Window>` is not implemented for `GnomeFactory`

error[E0277]: the trait bound `GnomeFactory: Factory<(dyn lib::elements::Button + 'static)>` is not satisfied
  --> $DIR/missing_interpolation.rs:18:6
   |
13 | #[abstract_factory(Factory, dyn Button, Window)]
   | ------------------------------------------------ required by this bound in `AbstractGuiFactory`
14 | pub trait AbstractGuiFactory {}
   |           ------------------ required by a bound in this
...
18 | impl AbstractGuiFactory for GnomeFactory {}
   |      ^^^^^^^^^^^^^^^^^^ the trait `Factory<(dyn lib::elements::Button + 'static)>` is not implemented for `GnomeFactory`
